### ISSUE TITLE
Unexpected behavior of property type guard

### ISSUE THREAD

**Issue by doronguttman** [microsoft/TypeScript#54031]
# Bug Report

### 🔎 Search Terms

type guard property 

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about type guards

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=5.0.4#code/JYOwLgpgTgZghgYwgAgBrIN4FgBQz-IwD2RA-AFzIDOYUoA5gNy4C+uuAJhAgDZxQoEREDWQAPSqmY5cQkWGRxKNOiHrIAvOIB0xItID0BgsgB6yCGIAO3SBwtQoRKJQAqATxvIA5CobIAH2QAVxAuGFAIDm9kYCpkECIFOCoqYHoQOAAjHhQwImQwTxRfWgZvbWQPL29Q8Mjo2PjE5NT0zJy8gqKavzUKgAoAJgBmIaGASnYZHGAYZAGxXRIJzFwTOVEs5TK1TR09aQ3hUQRJZAAyTEISHdUmZBZ9sSOCIxNzUMsbBDsHJxcVWKPlQMTiCSSijaGWyuUK3WB3nQVwwNyIdwYjEeFWQAAUnDYoEUfHowfEALZxNJ7UCFRGg5BZYIKAQAR2CwAE9lpPRKqL0GLUWJYg1G4ymeAImyIuW0PCI9AGWQANMgEBNpCwgA)

### 💻 Code

```ts
interface X {
    foo?: string;
}

declare const x: X;

const a: string = x.foo;
//    ^ expected error: Type 'string | undefined' is not assignable to type 'string'.
//                        Type 'undefined' is not assignable to type 'string'.(2322)


if (x.foo) {
    const b: string = x.foo;
    const c: X & { foo: string; } = x;
    //    ^ unexpected error: Type 'X' is not assignable to type 'X & { foo: string; }'.
    //                          Property 'foo' is missing in type 'X' but required in type '{ foo: string; }'.(2322)
    console.log(b, c);
}
```

### 🙁 Actual behavior

A type has an optional property. Using an `if` assertion statement, TypeScript knows to narrow the optional property when referring to the property itself, but does not know to "narrow" the object as having the property defined.

### 🙂 Expected behavior

Inside the `if` assertion, I would expect TypeScript to know that the object has the property defined.
